ntp-proto/src/algorithm/kalman/peer.rs

```diff
@@ -0,0 +1,1077 @@
+        packet: NtpPacket<'static>,
+    ) -> bool {
+        // Always keep the packet as last packet, since it reflects the most up-to-date
+        // information on the synchronization quality of the remote and any upcoming leap seconds.
+        self.last_packet = packet;
+
+        if measurement.localtime.is_before(self.filter_time) {
+            // Ignore the past
+            return false;
+        }
+
+        // Filter out one-time outliers (based on delay!)
+        if !self.prev_was_outlier
+            && (measurement.delay.to_seconds() - self.rtt_stats.mean())
+                > algo_config.delay_outlier_threshold * self.rtt_stats.variance().sqrt()
+        {
+            self.prev_was_outlier = true;
+            self.last_iter = measurement.localtime;
+            return false;
+        }
+
+        // Environment update
+        self.progress_filtertime(measurement.localtime);
+        self.rtt_stats.update(measurement.delay.to_seconds());
+
+        let (chi, weight, measurement_period) = self.absorb_measurement(measurement);
+
+        self.update_wander_estimate(algo_config, chi, weight);
+        self.update_desired_poll(config, algo_config, chi, weight, measurement_period);
+
+        debug!(
+            "peer offset {}+-{}ms, freq {}+-{}ppm",
+            self.state.entry(0) * 1000.,
+            (self.uncertainty.entry(0, 0) + sqr(self.last_packet.root_dispersion().to_seconds()))
+                .sqrt()
+                * 1000.,
+            self.state.entry(1) * 1e6,
+            self.uncertainty.entry(1, 1).sqrt() * 1e6
+        );
+
+        true
+    }
+
+    fn process_offset_steering(&mut self, steer: f64) {
+        self.state = self.state - Vector::new(steer, 0.0);
+        self.last_measurement.offset -= NtpDuration::from_seconds(steer);
+        self.last_measurement.localtime += NtpDuration::from_seconds(steer);
+        self.filter_time += NtpDuration::from_seconds(steer);
+    }
+
+    fn process_frequency_steering(&mut self, time: NtpTimestamp, steer: f64) {
+        self.progress_filtertime(time);
+        self.state = self.state - Vector::new(0.0, steer);
+        self.last_measurement.offset += NtpDuration::from_seconds(
+            steer * (time - self.last_measurement.localtime).to_seconds(),
+        );
+    }
+}
+
+#[derive(Debug, Clone)]
+#[allow(clippy::large_enum_variant)]
+enum PeerStateInner {
+    Initial(InitialPeerFilter),
+    Stable(PeerFilter),
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct PeerState(PeerStateInner);
+
+impl PeerState {
+    pub fn new() -> Self {
+        PeerState(PeerStateInner::Initial(InitialPeerFilter {
+            roundtriptime_stats: AveragingBuffer::default(),
+            init_offset: AveragingBuffer::default(),
+            samples: 0,
+        }))
+    }
+
+    pub fn update(
+        &mut self,
+        config: &SystemConfig,
+        algo_config: &AlgorithmConfig,
+        measurement: Measurement,
+        packet: NtpPacket<'static>,
+    ) -> bool {
```

Comment:
unclear from name and type: does this perform the update, or signal that an update is needed?

Reply:
Fixed